Speed up request buffered hashes

emhane opened the pull request:
closes https://github.com/paradigmxyz/reth/issues/6148. closes https://github.com/paradigmxyz/reth/issues/6308.

speeds up requesting buffered hashes by

~~- [x] dividing hashes store (`unknown_hashes`, `buffered_hashes`[ and `meta`]) into eth68 and eth66. this means only hashes will be traversed that can be included in the request being assembled in `fill_request_from_buffer_for_peer`.~~
- [x] improve performance and maintainability at cost of allocating memory for metadata for all hashes, regardless if seen in eth66 announcements only and hence have no metadata.
- [x] use the `transactions` list on `Peer` type to search `buffered_hashes` after `pop_any_idle_peer` returns. that cache has capacity 10 240 and buffered hashes has capacity 25 600. also, then we just have to search buffered hashes, and not nested lists (even if they just default to 3 elements long) in buffered hashes for our peer returned by `pop_any_idle_peer`. effectivity of this is up to how well we can update the `transactions` list in the `Peer` type **on-op**. ~~we will only update the `transactions` list on `Peer` type when we need to touch it anyway, **on-op**.~~ we won't move around elements in lists of peer's seen transactions. it serves only as a hint to which hash **cannot** be pending, not as a perfect list of which hashes are pending. this totally satisfies requirements and otherwise it doesn't scale.

mattsse commented:
this is still very expensive because we blindly iterate all hashes in the hope the peer is the fallback, can we not keep track of the peer's hashes instead when registering them as fallback peers?

emhane commented on crates/net/network/src/transactions/mod.rs:
```diff
@@ -1113,7 +1159,8 @@ where
 
         // try drain buffered transactions
-        this.request_buffered_hashes();
+        this.request_buffered_hashes::<TxHash68>();
+        this.request_buffered_hashes::<TxHash66>();
```
it does speed up things when the peer is sending both eth66 and eth68 messages. when assembling a request, eth66 uses hashes count to determine when the request is full. eth68 uses the size metadata to determine when a request is full. this means it doesn't make sense to mix eht68 and eth66 hashes in the same request. this change means that max half the number of hashes will be searched if the node has 50% hashes from eth66 announcements and 50% hashes form eth68 announcements. furthermore, it means only hashes that have the possibility to be included will be iterated over. if eth68 and eth66 hashes are in one, then for each hash it needs to search the eth68 metadata cache to find out if that is an eth66 or eht68 hash, and then the version which is not the same as the request being assembled will just be skipped.

emhane replied:
calling the suggested code twice only simplifies things, since mixed request with both versions eth66 and eth68 aren't assembled.

emhane replied:
a way to combine them, is discarding eth68 metadata when assembling a request from the buffer for some peer, so assembling the request as if were all hashes announced in an eth66 message. second option is to have a an average size guess for a tx with no size metadata, for example 512 bytes = 2 MiB / 4096, soft limit for max response size when assembling response / max number of hashes in an announcement. then we could still make use of the size metadata that exists which is nice.

emhane commented:
> this is still very expensive because we blindly iterate all hashes in the hope the peer is the fallback, can we not keep track of the peer's hashes instead when registering them as fallback peers?

we don't even need to register any peer as fallback peer in that case, it's just double up storage of same data that can be derived by checking peer's seen txns against buffered hashes (buffered as in buffered for re-fetch or for first fetch if the hash didn't fit in the request that was triggered by processing the announcement in which the hash was seen the first time).

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -36,42 +37,99 @@ const MARGINAL_FALLBACK_PEERS_PER_TX: u8 = 1;
+pub(super) const DEFAULT_MAX_CONCURRENT_TX_REQUESTS: u32 = 10000;
+
+/// Maximum concurrent [`GetPooledTxRequest`]s to allow per peer.
+pub(super) const MAX_CONCURRENT_TX_REQUESTS_PER_PEER: u8 = 1;
```
Is this not too low given the default max peer count is around 100?

emhane replied:
thanks for the input. then rather `DEFAULT_MAX_CONCURRENT_TX_REQUESTS` is too high and should also default to 100.

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -36,42 +37,99 @@ const MARGINAL_FALLBACK_PEERS_PER_TX: u8 = 1;
+    tx_fetcher_info: TransactionFetcherInfo,
+}
+
+#[derive(Debug)]
+pub(super) struct TxUnknownToPool {
```
nit: smol docs would be nice

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -101,203 +158,309 @@ impl TransactionFetcher {
+                break idle_peer.copied()
+            }
+
+            if let Some(bud) = budget {
+                budget = Some(bud.saturating_sub(1));
+                if budget == Some(0) {
+                    return None
+                }
+            }
```
nit

```suggestion
            if let Some(ref mut bud) = budget {
                *bud = bud.saturating_sub(1);
                if *bud == 0 {
                    return None
                }
            }
```

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -101,203 +158,309 @@ impl TransactionFetcher {
+    /// Returns left over hashes.
+    ///
+    /// Loops through hashes passed as parameter and checks if a hash fits in the expected
+    /// response. If no, it's add to surplus hashes. If yes, it's add to hashes to request and
```
nit 
```suggestion
    /// response. If no, it's added to surplus hashes. If yes, it's added to hashes to the request and
```

also, reading this fn over, i wasn't aware at first that it returned the surplus hashes, it would be nice to add this in the docs 😊

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -101,203 +158,309 @@ impl TransactionFetcher {
+            if size >= SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_MESSAGE {
+                return hashes_from_announcement_iter.collect::<RequestTxHashes>()
+            } else {
+                acc_size_response = size;
```
is this an accumulator? if so, is this supposed to be `acc_size_response += size`?

emhane replied:
this is not called in a loop. the loop is below.

onbjerg replied:
you're right

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -101,203 +158,309 @@ impl TransactionFetcher {
-                surplus_hashes.push(hash);
-                false
+    /// Tries to buffer hashes for retry. Hashes that have been re-requested
+    /// [`MAX_REQUEST_RETRIES_PER_TX_HASH`], are dropped.
```
is this check done indirectly?

emhane replied:
yepp. can move that doc line.

mattsse commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -36,42 +37,99 @@ const MARGINAL_FALLBACK_PEERS_PER_TX: u8 = 1;
 }
 
+#[derive(Debug)]
+pub(super) struct TxUnknownToPool {
```
this should be moved below the impl Block

mattsse commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -101,203 +158,311 @@ impl TransactionFetcher {
+        hashes_from_announcement: ValidAnnouncementData,
+    ) -> RequestTxHashes {
+        let (mut request_hashes, _version) = hashes_from_announcement.into_request_hashes();
+        if request_hashes.len() <= GET_POOLED_TRANSACTION_SOFT_LIMIT_NUM_HASHES {
```
flip arms

mattsse commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -36,42 +37,99 @@ const MARGINAL_FALLBACK_PEERS_PER_TX: u8 = 1;
-pub(super) struct TransactionFetcher {
-    /// All peers to which a request for pooled transactions is currently active. Maps 1-1 to
-    /// `inflight_requests`.
+pub(super) struct TransactionFetcher<T = TxUnknownToPool> {
```
is this generic actually useful?

mattsse reviewed (changes requested):
I think I can mostly follow along here, but the entire flow is not very obvious and needs to be properly documented.

My biggest concern with this is still that we're using the hashes as the main driver for creating requests and looping over all of them until the budget is exhausted.

it's not obvious what the difference between those two is

```rust
    /// Hashes that are awaiting an idle peer so they can be fetched.
    pub(super) hashes_pending_fetch: LruCache<TxHash>,
    /// Tracks all hashes in the transaction fetcher.
    pub(super) hashes_unknown_to_pool: LruMap<TxHash, T, Unlimited>,
```

besides one being bounded.


we previously had a lot of debug panics, so I'd like to have a bit more test coverage for this.
ideally also flamgraphing/samply a testnet node.

emhane commented:
> I think I can mostly follow along here, but the entire flow is not very obvious and needs to be properly documented.
> 

Will use the discussion below as a base for docs.

> My biggest concern with this is still that we're using the hashes as the main driver for creating requests and looping over all of them until the budget is exhausted.
> 

The reason for this, is that, when a node has been operating for a while, the cache in the type `Peer` called `transactions`, will be at capacity all the time. Transactions are never deleted from that cache. On the other hand, the pending hashes cache will vary in size depending on network conditions. This means that having a budget for how many of the peer's seen transactions to loop through, doesn't really make sense. Having a budget on how many of the peer's seen transactions to loop through instead would mean that the oldest hashes seen by the peer, never get a chance to be fetched from that peer if they are pending. How it is now, that the budget is on how many pending hashes to loop through, leaves a chance for even the _least_ recently used pending hashes to be fetched. This is because the pending hashes cache will be emptied of each hash that is fetched, so it gets smaller, meaning remaining hashes will be considered more recently used. Not having a budget is too risky, we need to be able to guarantee, that the operation of requesting pending hashes, will return within constant time when the node has high churn. TL;DR having a budget on how many of the transactions seen by the peer to loop through, doesn't make sense because that cache will eventually always be at capacity, whereas having a budget on how many pending hashes to loop through, makes sense because that cache shrinks when hashes are fetched.

Previously, it was mentioned that we should have a new mapping, `peer`->`pending-hashes-seen-by-peer`, for this purpose. Some questions and answers when considering introducing this new mapping here, make it clear why it isn't better:

When would we update it? 

Every time we buffer some hashes on behalf of the peer we would add hashes to the mapping. Like for example, it couldn't fetch them because it was busy or because they didn't fit into the request. The mapping would be a subset of the mapping `peer`->`transactions-seen-by-peer` discussed in the previous paragraph, that is already existent. 

So, when would we remove hashes from the mapping?

Every time a hash is successfully fetched. If that is the case, then we have the same problem we wanted to get away from originally because of time complexity: we would have to loop through all the mappings `peer`->`pending-hashes-seen-by-peer`, to check if the successfully fetched hash is in there and remove it. This would probably be more costly than the original problem of looping through fallback peers of all pending hashes, since the fallback peers cache had a capacity of just 3 peers, whereas this cache `pending-hashes-seen-by-peer` would have to have a far bigger capacity. Obviously we don't want that.

What do we do if we don't want to remove pending hashes upon successful fetch?

We leave it up to a bounded cache to do all the evictions.

What should the bound be on this cache?

We have no clue, it depends on how so many factors play together in the transaction fetcher and transaction manager code. So, we would make it big enough. This cache that we would have, would after the node has been running for a while, always be at capacity.

That's a lot of memory. For what price? How much efficiency in time complexity are we actually gaining here?

By the time we want to loop through the mapping for some peer, chances are high that some other peer has already fetched those hashes. So they are in fact not pending at all. This is the case since, the mempool is a gossip protocol, so redundancy of messages is a feature of the mempool. The possibility of keeping the mapping up to date, was out ruled, because it was too costly to loop through all the nested collections of pending hashes. So the gains in time complexity don't break even with the losses in space complexity.

We are absolutely loosing on space complexity, with the mapping `peer`->`pending-hashes-seen-by-peer`. Especially, considering that `pending-hashes-seen-by-peer` is a subset of the cache of all transactions seen by the peer, that is already available to use since before this pr, discussed in the first paragraph. This mapping that we already have, is an `LruCache`, which under the hood has a `LinkedHashSet`. Lookups in a hash set are O(1).

Additionally I implemented a separation of that cache into two, one containing hashes that can't possibly be pending and one containing hashes which have a chance of being pending, to decrease the likelihood of collisions. Hashes of transactions that are **sent to the peer** or have been received in full **over broadcast** from the peer, are added to the first. Hashes seen in an announcement are added to the second. This separation serves a decent decrease in chance of collision when looking up a hash in the peers seen hashes cache. It is **not** supposed to be a perfect separation of hashes that have been sent to or received from the peer in full. This means hashes are not moved from the `transactions_received_as_hash` cache and put in the `transactions_received_in_full_or_sent` cache when they are received in a response from fetching hashes - that would be too expensive for not much less chance of collision.

What I have implemented in terms of flow:
(i) Looping through the lru pending hashes, looking up the fallback peers for each hash, and checking if those peers are idle. On the **first** peer that is idle, **stop looping**. The likelihood that the lru or the second to lru pending hash has an idle fallback peer, is high. The fallback peers cache, has a capacity of 3 peers. I don't think this will be a problematic in terms of time complexity at all.

(ii) There is a peer from (i) along with one pending hash to request from it. Now the request will be filled up with more pending hashes. Loop through the pending hashes in lru order again, until a budget is depleted or the request is full. For each hash, look it up in the peers seen transactions mapping, the one for 'hashes that could be pending', `transactions_received_as_hash`. If it contains the hash, add it and update the total size.  

> it's not obvious what the difference between those two is
> 
> ```rust
>     /// Hashes that are awaiting an idle peer so they can be fetched.
>     pub(super) hashes_pending_fetch: LruCache<TxHash>,
>     /// Tracks all hashes in the transaction fetcher.
>     pub(super) hashes_unknown_to_pool: LruMap<TxHash, T, Unlimited>,
> ```
> 
> besides one being bounded.
> 

That is unintentional! They are both supposed to be bounded. I was thinking about what the bounds should be. The original bound, before the separation into two caches, was `10 * 1024`. After some consideration, I increased them to `10 * 1024`, but having both at `8 * 1024` should probably be fine. Adding metrics will be able to say after some time which portion of the original `10 * 1024` bytes, each holds. They sets of hashes in the two caches are disjoint.

> we previously had a lot of debug panics, so I'd like to have a bit more test coverage for this. ideally also flamgraphing/samply a testnet node.

absolutely.

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -471,180 +684,121 @@ impl TransactionFetcher {
-            acc_size_response, hashes, self
-        );
+        // if request full enough already, we're satisfied, send request for single tx
+        if acc_size_response >= SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_FETCH_PENDING {
```
if accumulated size of the response is as big as the limit, there isn't any more space to fill

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -471,180 +684,121 @@ impl TransactionFetcher {
+            // 4. Check if acc size or hashes count is at limit, if so stop looping.
+            // if request full enough or the number of hashes in the request is enough, we're
+            // satisfied
+            if acc_size_response >=
```
same here, if the expected response is at the limit, it can't be filled anymore, and is a good criteria to return.

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -471,180 +684,121 @@ impl TransactionFetcher {
-        for hash in hashes {
-            self.buffered_hashes.remove(hash);
+        // 5. Remove hashes to request from cache of hashes pending fetch.
+        for hash in hashes_to_request.iter() {
+            self.hashes_pending_fetch.remove(hash);
```
the first hash hash in the request buffer `hashes_to_request`, has already been removed here
https://github.com/paradigmxyz/reth/blob/2afeb89ecbe937d3b2af20dc393311b788f1ecee/crates/net/network/src/transactions/fetcher.rs#L253

mattsse commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -101,203 +186,311 @@ impl TransactionFetcher {
+    ///
+    /// Loops through the hashes pending fetch in lru order until one is found with an idle
+    /// fallback peer, or the budget passed as parameter is depleted, whatever happens first.
+    pub(super) fn find_any_idle_fallback_peer_for_any_pending_hash(
```
ty, this is very clear :)

mattsse commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -17,61 +21,143 @@ use tokio::sync::{mpsc::error::TrySendError, oneshot, oneshot::error::RecvError}
-    /// request retries and last recently seen fallback peers (max one request try for any peer).
-    pub(super) unknown_hashes: LruMap<TxHash, (u8, LruCache<PeerId>), Unlimited>,
-    /// Size metadata for unknown eth68 hashes.
-    pub(super) eth68_meta: LruMap<TxHash, usize, Unlimited>,
+    pub(super) hashes_pending_fetch: LruCache<TxHash>,
+    /// Tracks all hashes in the transaction fetcher.
+    pub(super) hashes_unknown_to_pool: LruMap<TxHash, T, Unlimited>,
```
it's not obvious what the difference here is.

> Hashes that are awaiting an idle peer so they can be fetched.

> Tracks all hashes in the transaction fetcher.

aren't those the same thing?

emhane replied:
from compiler development there is a very nice type in the [cranelift compiler called `PrimaryMap`](https://github.com/bytecodealliance/wasmtime/blob/2c0e528b9fad9b93e55b03abb816360f3c1bb52d/cranelift/entity/src/primary.rs#L16-L39). `PrimaryMap` is a dense index space in the sense that [all keys are sequential and are created upon pushing a value into the map](https://github.com/bytecodealliance/wasmtime/blob/de051941c98e30eb86637206fd50e3599eba0853/cranelift/entity/src/primary.rs#L121-L126). then there is a another type called [`SecondaryMap`](https://github.com/bytecodealliance/wasmtime/blob/de051941c98e30eb86637206fd50e3599eba0853/cranelift/entity/src/map.rs#L18-L35), which doesn't allocate entities, but rather uses the entities from the `PrimaryMap` as keys. It's also enforced that there can be max one `PrimaryMap` instance for a type that implements `EntityRef`. this is the pattern I wanted to follow here, `hashes_unknown_to_pool` is the "primary map" and `hashes_pending_fetch` is the "secondary map".

since we can't fit a 256-byte `TxHash` in a `usize`, we can't implement `EntityRef` for it. I wanted to use some similar types here, but didn't find a simple way. this is the case since other integrated components, like for example a session that is forwarding transactions to the tx fetcher, uses the `TxHash` to index transactions, and they all need to use the same index to avoid expensive look ups for translating one components index to another components index when interfacing between components. so using a smaller type that implements EntityRef like `u32`  to index hashes, would be a crate-wide change, maybe even workspace-wide change. it would be nice but is out of scope of this pr. then we would have a map somewhere with a `u32->TxHash` mapping, and only pass the 32-bit reference around instead of the 256-bit reference we do now. but! we can't do that since the hashes in circulation in the node always change. we can do `u32->TxHashContainer` in a fixed size store for hashes.

the store `hashes_unknown_to_pool` should be flexible sized for now, so that's off the table here. will think of better names for the fields though.

mattsse reviewed (commented):
thanks for the detailed write up.

I found this a lot easier to follow now.

what's not immediately obvious is how the two caches differ, or rather how we keep track of hashes that are currently inflight.

emhane commented:
This is how the branch is looking on mainnet rn.

<img width="1414" alt="Screenshot 2024-02-11 at 00 31 26" src="https://github.com/paradigmxyz/reth/assets/58548332/e1297932-6bbf-403a-9421-5291d79387af">
<img width="1408" alt="Screenshot 2024-02-11 at 00 31 36" src="https://github.com/paradigmxyz/reth/assets/58548332/58ce3341-6bf1-4af1-a4d9-9af9d88fb281">

This is simply at commit [739d3c1](https://github.com/paradigmxyz/reth/pull/6318/commits/739d3c1d89b4a9a285b0da3f880ba1a0a77b8d3f) 
https://share.firefox.dev/3STlLDb

Will update with new samply from latest commit later.

Rjected commented on crates/net/network/src/transactions/constants.rs:
```diff
@@ -0,0 +1,201 @@
+
+    /* ====== SCALARS FOR USE ON FETCH PENDING HASHES ====== */
+
+    /// Default soft limit for the number of hashes in a
+    /// [`GetPooledTransactions`](reth_eth_wire::GetPooledTransactions) request, when it is filled
+    /// from hashes pending fetch. Default is half of the
+    /// [`SOFT_LIMIT_COUNT_HASHES_IN_GET_POOLED_TRANSACTIONS_REQUEST`] which by spec is 256
+    /// hashes, so 128 hashes.
+    pub const DEFAULT_SOFT_LIMIT_COUNT_HASHES_IN_GET_POOLED_TRANSACTIONS_REQUEST_ON_FETCH_PENDING_HASHES:
+    usize = SOFT_LIMIT_COUNT_HASHES_IN_GET_POOLED_TRANSACTIONS_REQUEST / 2;
+
+    /// Default soft limit for a [`PooledTransactions`](reth_eth_wire::PooledTransactions) response
+    /// when it's used as expected response in calibrating the filling of a
+    /// [`GetPooledTransactions`](reth_eth_wire::GetPooledTransactions) request, when the request
+    /// is filled from hashes pending fetch. Default is half of
+    /// [`DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_PACK_GET_POOLED_TRANSACTIONS_REQUEST`],
+    /// which defaults to 128 KiB, so 64 KiB.
+    pub const DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_FETCH_PENDING_HASHES:
+        usize = DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_PACK_GET_POOLED_TRANSACTIONS_REQUEST / 2;
```
I like the documentation on these but these names are gigantic, is there any way to make this more concise?

emhane replied:
yeah I know they are, but not having enough info can cost a lot of time since then they are interpreted wrong. we already had this with some other constants recently. will eventually change some of these to const functions anyway, then we can asses the lengths again.

Rjected reviewed (commented):
Will do a more comprehensive review soon, but I just saw these constant names and wanted to see if they could be made shorter

onbjerg reviewed (approved):
lgtm

mattsse reviewed (approved):
sharing @Rjected view that some names are gigantic, but can bikeshed separately.

lgtm, nice work
